apis/r/R/SOMASparseNDArray.R

```diff
@@ -189,6 +189,30 @@ SOMASparseNDArray <- R6::R6Class(
     #' @return A scalar with the number of non-zero elements
     nnz = function() {
       nnz(self$uri, config=as.character(tiledb::config(self$tiledbsoma_ctx$context())))
+    },
+
+    #' @description Increases the shape of the array as specfied. Raises an error
+    #' if the new shape is less than the current shape in any dimension. Raises
+    #' an error if the new shape exceeds maxshape in any dimension. Raises an
+    #' error if the array doesn't already have a shape: in that case please call
+    #' tiledbsoma_upgrade_shape.
+    #' @return No return value
+    resize = function(newshape) {
+      # TODO: move this to SOMANDArrayBase.R once core offers current-domain support for dense arrays.
+      # https://github.com/single-cell-data/TileDB-SOMA/issues/2955
+      newshape <- unlist(newshape, use.names = FALSE)
+      resize(self$uri, newshape, config=as.character(tiledb::config(self$tiledbsoma_ctx$context())))
+    },
+
+    #' @description Allows the array to have a resizeable shape as described in the
+    #' TileDB-SOMA 1.15 release notes.  Raises an error if the new shape exceeds maxshape in any
+    #' dimension. Raises an error if the array already has a shape.
+    #' @return No return value
+    tiledbsoma_upgrade_shape = function(newshape) {
```

Comment:
What is the workflow for going from non-resizeable to resizeable? I would assume that the order would be:
- open old array
- request an upgrade to resizeable array
- resize

But it seems like we can do
- open an old array
- request an upgrade to resizeable _and_ resize at the same time

Was this second workflow explictly agreed to? Also, do we want to raise an error if the array is already resizeable? Is this error classed so that we can catch it and move on à la
```r
array <- SOMASparseNDArrayOpen(uri)
tryCatch(array$tiledbsoma_upgrade_shape(), alreadyResizeableError = \(...) NULL)
array$resize(new_shape)
```

Reply:
> Was this second workflow explictly agreed to? 

@mojaveazure if you mean what I think you mean, it's not only not agreed to, it's not implemented

* `tiledbsoma_upgrade_shape` throws if the array has a shape
* `resize` throws if it does not

Reply:
My question was do we want `tiledbsoma_upgrade_shape()` to make arrays resizeable _and_ perform a resize at the same time, or do we want these to be separate steps?

Reply:
The spec discussion was we wanted the latter -- separate steps. And I don't want to re-litigate spec discussions.

And the former is not possible as I've got it coded.

So I'm with the spec. I'll make this clearer in the comments.

Reply:
@mojaveazure ah, I think our terminology might be not quite aligned.

* The core domain (soma max shape) is immutable. If it's set to 100 x 200 then it can never be set higher.
  * Every TileDB array already has a core domain -- you can't have an array without one
* A TileDB array can have a core current domain (soma shape) -- old arrays won't have one; new arrays will
* If the array has no core current domain (soma shape), a current domain (soma shape)can be set on it. This is what `tiledbsoma_upgrade_shape` does.
  * The core current domain (soma shape) being set must be `<=` the core domain (soma maxshape) in every slot.
  * So if the core domain is 100x200 then the current domain can be set to 20 x 30, or 100 x 200, but it can't be set to 1000 x 2000.
* If the array has has a core current domain (soma shape), it can be changed:
  * The new core current domain (new soma shape) must be >= the old current domain (old soma shape)
  * The new core current domain (new soma shape) must be <= the core domain (soma maxshape)

So to "make an array resizable":

* If the existing maxshape is 100 x 200 then that is immutable and there is no facility to change that, other than re-creating the array
* If the array doesn't have soma shape, one can be added
* If the array does have one, it can be extended, up to the maxshape

Reply:
Phrased even more compactly:

* What we're now calling the `maxshape` (the core domain) was and is immutable
* The new `shape` is a new thing which can be _added_, which, if present, can be smaller than the `maxshape`, and this _new_ thing can be resized up to the `maxshape`

Reply:
Okay, IIUC `$tiledbsoma_upgrade_shape()` adds a new SOMA shape to the array, and `$resize()` expands the SOMA shape to `soma_shape` <= `new_shape` <= `array_domain`? In that case, the name `new_shape` for `$tiledbsoma_upgrade_shape()` is confusing as it implies (to me and other biologists) that this method is doing "add SOMA shape" _and_ resize at the same time as its the same parameter name as `$resize()`. I would call it something like `shape` or `soma_shape` to differentiate it from `$resize()`